### ISSUE TITLE
[bug] Different package ids in lockfiles for some packages, depending on default_package_id_mode 

### ISSUE THREAD

**Issue by a4z** [conan-io/conan#8424]
### Environment Details (include every applicable attribute)
  * Operating System+version: Mac OSX
  * Compiler+version: not relevant
  * Conan version: 1.33
  * Python version: 3.8.7

### Steps to reproduce (Include if Applicable)

same environment, I generate lockfiles for the same conanfile with the same profiles ... 

```
conan config set general.default_package_id_mode=full_version_mode
```
gives me this

```
    alljoyn/0.16.10:f83037eff23ab3a94190d7f3f7b37a2d6d522241 - Download
    catch2/2.13.1:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    cpp-httplib/0.7.8:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    djinni-support-lib/0.0.1:a9473fb39d33210ba7a93e68c439b8e32b48cf5c - Download
    doctest/2.4.0:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    fmt/7.0.3:6a83d7f783e7ee89a83cf2fe72b5f5f67538e2a6 - Download
    libcurl/7.74.0:10fc4e3881da1c7f6c5f8010fc987baaf18af2f2 - Download
    lua/5.3.5:97560902203b831a69203233a5d77d092aae1af6 - Download
    magic_enum/0.6.6:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    miniz/2.1.0:647afeb69d3b0a2d3d316e80b24d38c714cc6900 - Download
    ms-gsl/2.0.0:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    nlohmann_json/3.9.1:9c8f7bbf5ecfc34e9fa77921cffb72b1ad4f2325 - Download
    openssl/1.1.1i:647afeb69d3b0a2d3d316e80b24d38c714cc6900 - Download
    paho-mqtt-c/1.3.8:c0e6b2a0b169ea50c3ed9c3ac7f2ea9f2263e4b5 - Download
    sol2/3.2.1:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    zlib/1.2.11:647afeb69d3b0a2d3d316e80b24d38c714cc6900 - Download
```

and now 
```
conan config set general.default_package_id_mode=recipe_revision_mode
```
the exactly same command to produce a lockfile gives me 
```
    alljoyn/0.16.10:f83037eff23ab3a94190d7f3f7b37a2d6d522241 - Download
    catch2/2.13.1:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    cpp-httplib/0.7.8:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    djinni-support-lib/0.0.1:a9473fb39d33210ba7a93e68c439b8e32b48cf5c - Download
    doctest/2.4.0:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    fmt/7.0.3:6a83d7f783e7ee89a83cf2fe72b5f5f67538e2a6 - Download
    libcurl/7.74.0:201636c1f31b8ff949b47d563cd024dce2ba08b5 - Download
    lua/5.3.5:97560902203b831a69203233a5d77d092aae1af6 - Download
    magic_enum/0.6.6:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    miniz/2.1.0:647afeb69d3b0a2d3d316e80b24d38c714cc6900 - Download
    ms-gsl/2.0.0:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    nlohmann_json/3.9.1:9c8f7bbf5ecfc34e9fa77921cffb72b1ad4f2325 - Download
    openssl/1.1.1i:647afeb69d3b0a2d3d316e80b24d38c714cc6900 - Download
    paho-mqtt-c/1.3.8:bbb709a4259669b6e55572c812ba3acb4ede08d6 - Download
    sol2/3.2.1:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Download
    zlib/1.2.11:647afeb69d3b0a2d3d316e80b24d38c714cc6900 - Download
```

you might notice the difference, but I highlight it anyway

```
   libcurl/7.74.0:10fc4e3881da1c7f6c5f8010fc987baaf18af2f2 - Download
   paho-mqtt-c/1.3.8:c0e6b2a0b169ea50c3ed9c3ac7f2ea9f2263e4b5 - Download
```

vs 

```
   libcurl/7.74.0:201636c1f31b8ff949b47d563cd024dce2ba08b5 - Download
   paho-mqtt-c/1.3.8:bbb709a4259669b6e55572c812ba3acb4ede08d6 - Download
```

this is unexpected. 

either all should differ, or none . Or do I miss something ?

**Comment by solvingj** [comment]
It's a really good question and I will investigate it more closely asap.  However, my first instinct is that this is actually a reasonable result. Here is the list of things which might be factors and which I need to think and test around:
- Perhaps it depends on which revisions are in your local cache
    - try running with clean cache and see if results are same)
- Perhaps these particular recipes have multiple revisions and the rest don't
- Perhaps these are the only transitive dependencies
- Perhaps these are the only direct dependencies which have other transitive dependencies
Hope to followup tomorrow, but maybe theses clues help you figure it out first.

**Comment by a4z** [comment]
1) both runs where done with a clean cache
2) this is possible, but I'd expect the latest to be returned in both cases (which is what I get via revision_mode) 
3) No, these are fist class dependencies
4) They have dependencies to the same packages, openssl and therefor zlib

**Comment by solvingj** [comment]
I have not yet had a chance to dissect the `package_id` calculation for each mode for these two packages to find a reason.  However, it seems likely to me that there is some conditional logic in these recipes which causes them to produce different `package_id` under different `default_package_id_mode` settings.  There are a number of legitimate conditional operations in multiple methods within recipes in which can have this result.  It seems very unlikely to be a bug.

**Comment by a4z** [comment]
if you not define it as a bug that's ok to me, 
but I am not sure what build/packag `full_version_mode ` would have actually given me, for sure not that one I expected.

Btw, it can it be that libtool has the same behaviour.
It's a build requirement, so I do not monitor that so intensive, but I think I saw that popping up in out logs.

but anyhow, I have set all our configs to `recipe_revision_mode` , I just noticed the problem by accident, since some container was not built as expected, and I was happy about the different id because that showed me the config in the container was wrong :-) 

but I will not observe that issue a lot in future since it affects me not that much. 
I just thought you would like to know about that since I still find it surprising, getting those different ids

**Comment by solvingj** [comment]
Yes, I would say that in general, switching between different `package_id` modes can result in different `package_ids` being calculated and the reasons and calculations can be un-intuitive at times.  Each time I have drilled down into similar cases for myself where the `package_id` behavior did something unexpected, I found there was a perfectly legitimate reason and was reminded just how complex conditional logic for options and settings in recipes make predicting the results.

If you come across an unexpected result in the future, and it actually impacts you long-term, lets dig into it at that time.
